main.c

```diff
@@ -417,25 +435,47 @@ static void poll_loop(const poll_loop_args_t* args)
                 args->mem_kill_percent, args->swap_kill_percent);
-        } else if (sig == SIGTERM) {
+        } else if (sig == SIGTERM && sigterm_countdown_ms == args->sigterm_delay_ms) {
             print_mem_stats(warn, m);
             warn("low memory! at or below SIGTERM limits: mem " PRIPCT ", swap " PRIPCT "\n",
                 args->mem_term_percent, args->swap_term_percent);
+
+            if(args->sigterm_delay_ms) {
+                // kickstart the countdown
+                victim = find_largest_process(args);
+                notify("earlyoom",
+                       "Low memory! Will send SIGTERM, likely to %s (%d), "
+                       "in %g or more seconds unless situation improves",
+                       victim.name, victim.pid,
+                       (double)args->sigterm_delay_ms / 1000);
+                sigterm_countdown_ms -= 1;
+            }
         }
-        if (sig) {
-            procinfo_t victim = find_largest_process(args);
-            /* The run time of find_largest_process is proportional to the number
-             * of processes, and takes 2.5ms on my box with a running Gnome desktop (try "make bench").
-             * This is long enough that the situation may have changed in the meantime,
-             * so we double-check if we still need to kill anything.
-             * The run time of parse_meminfo is only 6us on my box and independent of the number
-             * of processes (try "make bench").
-             */
-            m = parse_meminfo();
-            if (lowmem_sig(args, &m) == 0) {
-                warn("memory situation has recovered while selecting victim\n");
+
+        if (sig && (sig != SIGTERM || sigterm_countdown_ms <= 0)) {
+            victim = find_largest_process(args);
+
+            if (!args->sigterm_delay_ms) {
+                /* The run time of find_largest_process is proportional to the number
+                 * of processes, and takes 2.5ms on my box with a running Gnome desktop (try "make bench").
+                 * This is long enough that the situation may have changed in the meantime,
+                 * so we double-check if we still need to kill anything.
+                 * The run time of parse_meminfo is only 6us on my box and independent of the number
+                 * of processes (try "make bench").
+                 */
+                m = parse_meminfo();
+                if (lowmem_sig(args, &m) == 0) {
+                    warn("memory situation has recovered while selecting victim\n");
+                } else {
+                    kill_process(args, sig, victim);
+                }
             } else {
-                kill_process(args, sig, victim);
+                // given that SIGTERM window can be passed rather quickly,
+                // make one small attempt to terminate process gracefully;
+                // kill_wait() will quickly escalate to SIGKILL anyway
+                kill_process(args, SIGTERM, victim);
```

Comment:
Sorry, I glitched out and have had misread `kill_wait()`: this does not really make much sense. This line should be either just `kill_process(args, sig, victim);` or I'll possibly add a single adaptive sleep to not dip too deep into SIGKILL territory. I'm a bit busy now, so it might take a while on my end.